### PR TITLE
sandstone_ssl: Add our own random generator

### PR DESCRIPTION
**Pull request by AthenasJimenez**
This depends on PR #381.

### CODE REVIEW

File: framework/sandstone_ssl_rand.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    return 1;
+}
+
+// TODO: Add clean up funcitons
+void sandstone_ssl_rand_init()
+{
+    if (s_OSSL_PROVIDER_add_builtin(NULL, "sandstone-rand", sandstone_ssl_rand_provider_init) != 1
+            || s_RAND_set_DRBG_type(NULL, "sand-rand", NULL, NULL, NULL) != 1
+            || (s_prov = s_OSSL_PROVIDER_try_load(NULL, "sandstone-rand", 1)) == NULL )
+        printf("WARNING: Cannot load sandstone random provider");
```
Comment (thiagomacieira) [inline framework/sandstone_ssl_rand.cpp:159]: `fprintf` to `stderr`; please start the warning with "%s" / `program_invocation_name`.

File: framework/sandstone.cpp [hunk 2]
```diff
@@ -3755,7 +3756,10 @@ int main(int argc, char **argv)
 
 #if SANDSTONE_SSL_BUILD
     if (SANDSTONE_SSL_LINKED || sApp->current_fork_mode() != SandstoneApplication::exec_each_test)
+    {
```
Comment (thiagomacieira) [inline framework/sandstone.cpp:3759]: Nitpick: `{` to the previous line

File: framework/sandstone_ssl_rand.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STATE);
+    if (p != NULL && !s_OSSL_PARAM_set_int(p, *(int *)ctx))
+        return 0;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STRENGTH);
+    if (p != NULL && !s_OSSL_PARAM_set_int(p, 500))
+        return 0;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_MAX_REQUEST);
```
Comment (thiagomacieira) [inline framework/sandstone_ssl_rand.cpp:92]: What's the max request? Is that data length or the maximum value we return?

File: framework/sandstone_ssl_rand.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+static int sandstone_rand_generate(void *ctx, unsigned char *out, size_t outlen,
+                            unsigned int strength, int prediction_resistance,
+                            const unsigned char *addin, size_t addin_len)
+{
+    memset_random(out, outlen);
+
+    return 1;
+}
+
+// TODO: Do something
```
Comment (thiagomacieira) [inline framework/sandstone_ssl_rand.cpp:53]: We don't support reseeding. You can pretend to OpenSSL that we are a true RNG (that is, not a DPRNG).

**Review by thiagomacieira — COMMENTED** [review]
Holy indirection, Batman!

So we register a provider whose only content is an initialisation function. That function produces a new context whose only content is a dispatch table, adding a query function. The query function supports one query only and that returns another table, which points to yet another function table containing the actual operation functions!

Whew! Good work so far!

File: framework/sandstone_ssl_rand.cpp (hunk 3 above)
Comment (AthenasJimenez) [reply framework/sandstone_ssl_rand.cpp:92]: Good question, from documentation: "Specifies the maximum number of bytes that can be generated in a single call to OSSL_FUNC_rand_generate".  Do we have a limit for memset_random?

File: framework/sandstone_ssl_rand.cpp (hunk 4 above)
Comment (AthenasJimenez) [reply framework/sandstone_ssl_rand.cpp]: It's ok, it is not mandatory to support it. So I removed it.

File: framework/sandstone_ssl_rand.cpp (hunk 3 above)
Comment (thiagomacieira) [reply framework/sandstone_ssl_rand.cpp:84]: We don't.

**Review by thiagomacieira — APPROVED** [review]

**Pull request merged**